bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87

```diff
@@ -0,0 +1,152 @@
+[experiment]
+end_date = '2020-04-13'
+
+segments = ["attributed", "non_attributed"]
+
+enrollment_period = 8
+
+reference_branch = "baseline-newtab-content"
+
+
+## data sources
+
+[data_sources]
+
+## data source: AS session pings data
+[data_sources.as_session_ping]
+from_expression = "(SELECT client_id, user_prefs, date(submission_timestamp) as submission_date FROM  `moz-fx-data-shared-prod.activity_stream.sessions`)"
+
+experiments_column_type = "native"
+
+
+
+
+## Segments
+
+[segments]
+
+[segments.attributed]
+select_expression = 'attr_source is not null'
+data_source = "npp"
+
+[segments.non_attributed]
+select_expression = 'attr_source is null'
+data_source = "npp"
+
+[segments.data_sources.npp]
+from_expression = """
+    (SELECT
+        DATE(submission_timestamp) AS submission_date,
+        client_id,
+        environment.settings.attribution.source AS attr_source
+    FROM `moz-fx-data-shared-prod.telemetry.new_profile`
+    WHERE DATE(submission_timestamp) BETWEEN "{{experiment.start_date_str}}" AND "{{experiment.last_enrollment_date_str}}"
+    AND environment.partner.distribution_id IS NULL
+    AND coalesce(environment.settings.attribution.ua, '') != 'firefox'
+    )
+    """
+
+
+## metrics
+
+[metrics]
+overall = ['days_of_use', 'search_count', 'organic_search_count', 'unenroll', 
+           'pocket_rec_clicks', 'pocket_spoc_clicks', 'homepage_search_count', 
+           'topsite_clicks', 'highlight_clicks', 'homepage_switch', 
+           'turned_off_search', 'turned_off_topsites', 'turned_off_pocket', 
+           'turned_off_highlights']
+
+
+## homepage_search_count
+
+[metrics.homepage_search_count]
+select_expression = "SUM(CASE WHEN source IN ('newtab','abouthome') THEN sap ELSE 0 END)"
+data_source = 'search_clients_daily'
+
+[metrics.homepage_search_count.statistics.bootstrap_mean]
+
+
+## topsite_clicks
+
+[metrics.topsite_clicks]
+select_expression = "COUNTIF(
+                event = 'CLICK'
+                AND source = 'TOP_SITES')"
+data_source = 'activity_stream_events'
+
+[metrics.topsite_clicks.statistics.bootstrap_mean]
+
+
+## highlight_clicks
+
+[metrics.highlight_clicks]
+select_expression = "COUNTIF(
+                event = 'CLICK'
+                AND source = 'HIGHLIGHTS')"
+data_source = 'activity_stream_events'
+
+[metrics.highlight_clicks.statistics.bootstrap_mean]
+
+
+## homepage_switch
+    # NOTE: this metric doesn't take into account users setting their page to 
+    # blank. Only when users have set the page to a 3rd party! 
+
+[metrics.homepage_switch]
+select_expression = "MAX(
```

Comment:
If a client never sends an activity stream event ping this will evaluate to NULL and the stat will choke. You should either a) write something like COALESCE(MAX(...), FALSE) to provide a default value, or use `pre_treatments = ["drop_nulls"]` on the statistic declaration if it doesn't make sense to include those users in the calculation in any way.

Reply:
ah, I see. I do want to provide a default value here. 

So I'm guessing the select expression is evaluating after the enrollment table has been joined to the data source, correct? 

+1 on the heads up. 

Hot take, the COALESCE solution works, but feels a bit like a hack-y (and requires a bit of mozanalysis plumbing context). Would be good to call this out as a common use case, or maybe parameterize it like we pre_treatments. 

Specificially, I mean, "how do you want to deal with values when there's isn't a match on the join".

Reply:
The premise, which I think is basically right, is that the user must always make a decision about how to handle NULLs. Right now the metrics definitions are a thin wrapper around mozanalysis metrics, which spit the select_expression directly into a query.

An alternative and maybe more elegant user interface could be something like adding a mandatory `if_null` key to every metric definition that accepts either a value or e.g. `"drop"`. That at least brings "you forgot to handle nulls" errors forward to the point of definition, which I like a lot.

But the Jetstream Metric object will have to diverge a little from the mozanalysis Metric in that case because mozanalysis can't actually drop nulls in the results dataframes that it emits, because there's one row per _client_ and we definitely don't want to remove that client from other metrics just because they had a null for one of them. That's why we've associated the drop_nulls pretreatment with statistics, which operate columnwise on a single metric.

And because the Jetstream metrics diverge from the mozanalysis metrics, defining a metric in mozanalysis will no longer be sufficient for a metric to be usable with Jetstream, because we need to supply the additional context somewhere about how to handle nulls. This is already sort of true because we have to provide default statistics for them. Maybe mozanalysis metrics get a special warrant that the null handling is correct, idk.

tl;dr there are some tradeoffs, but I agree this is clunky, so maybe they're worth it.